core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -251,4 +322,126 @@ private static void validatePlanningArguments(long splitSize, int lookback, long
+          CloseableIterable<T> splitFiles,
+          long splitSize,
+          int lookback,
+          Function<T, Long> weightFunc,
+          Function<List<T>, G> groupFunc) {
+
+    return CloseableIterable.transform(
+        CloseableIterable.combine(
+            new BinPacking.PackingIterable<>(splitFiles, splitSize, lookback, weightFunc, true),
+            splitFiles),
+        groupFunc);
+  }
+
+  private static class AdaptiveSplitPlanningIterable<T extends ScanTask, G extends ScanTaskGroup<T>>
+      extends CloseableGroup implements CloseableIterable<G> {
+    private final CloseableIterable<T> files;
+    private final int parallelism;
+    private final long splitSize;
+    private final int lookback;
+    private final Function<T, Long> weightFunc;
+    private final BiFunction<CloseableIterable<T>, Long, CloseableIterable<T>> splitFunc;
+    private final Function<List<T>, G> groupFunc;
+
+    private Long targetSize = null;
+
+    private AdaptiveSplitPlanningIterable(
+        CloseableIterable<T> files,
+        int parallelism,
+        long splitSize,
+        int lookback,
+        Function<T, Long> weightFunc,
+        BiFunction<CloseableIterable<T>, Long, CloseableIterable<T>> splitFunc,
+        Function<List<T>, G> groupFunc) {
+      this.files = files;
+      this.parallelism = parallelism;
+      this.splitSize = splitSize;
+      this.lookback = lookback;
+      this.weightFunc = weightFunc;
+      this.splitFunc = splitFunc;
+      this.groupFunc = groupFunc;
+    }
+
+    @Override
+    public CloseableIterator<G> iterator() {
+      if (targetSize != null) {
+        // target size is already known so plan with the static target size
+        CloseableIterable<T> splitTasks = splitFunc.apply(files, targetSize);
+        CloseableIterator<G> iter =
+            planTasksInternal(splitTasks, targetSize, lookback, weightFunc, groupFunc).iterator();
+        addCloseable(iter);
+        return iter;
+      }
+
+      boolean shouldClose = true;
+      CloseableIterator<T> tasksIter = files.iterator();
+      try {
+        // load tasks until the iterator is exhausted or until the total weight is enough to get the
+        // parallelism at the split size passed in.
+        LinkedList<T> readAheadTasks = Lists.newLinkedList();
+        long readToSize = parallelism * splitSize;
+        long totalSize = 0L;
+
+        while (tasksIter.hasNext()) {
+          T task = tasksIter.next();
+          readAheadTasks.addLast(task);
+          totalSize += weightFunc.apply(task);
+
+          if (totalSize > readToSize) {
+            break;
+          }
+        }
+
+        // if total size was reached, then the requested split size is used. otherwise, the iterator
+        // was exhausted and the split size will be adjusted to target parallelism with a reasonable
+        // minimum.
+        this.targetSize = Math.max(MIN_SPLIT_SIZE, Math.min(totalSize / parallelism, splitSize));
```

Comment:
This is where the split size is determined.

Reply:
I assume your intent is that the `splitSize` would be determined before calling adaptive planning to resolve whether it's configured at the table level or the split default.  The only concern is that we still need to incorporate other configurations on the table (like row group / block size) since that directly affects how we can split.

Reply:
This works by targeting the split size that was requested, but trying to maximize parallelism if there isn't enough data to reach that split size. The logic for the requested split size is the same as what we currently use.

The row groups size is still orthogonal to this. We don't need to change the split size based on the target row group size. That will happen automatically because we track the split offsets and will just not be able to split files any further than their existing row groups.

Reply:
Can we move this logic into a separate utility method? I think we will use that directly in Spark as we have a fully loaded list of tasks and there is no point in constructing `CloseableIterable` and complicating the logic in that case.

I think this approach would perform well for cases where the provided split size prevents us from utilizing all cores in the cluster. However, I'd also target the following two use cases (I am still thinking about the second one):

- The size of data to be scanned is large and the default split size produces too many tasks. That's the original issue the PR with a SQL property was trying to solve. The need to keep all that extra task info can easily kill the driver, not to mention the overhead of launching a task in Spark (easily a second).
- I think our goal is to produce a number of splits that is a multiple of the number of slots in the cluster. Suppose we produce 1100 tasks for 1000 cores that are all equally balanced. Spark will run first 1000 tasks in parallel (let's assume each task would take the same amount of time) and only after that schedule the remaining 100 tasks. The remaining 100 tasks would take the same amount of time as the first 1000 as most of the slots in the cluster will be idle. Instead of blinding following the configured split size, it would be much better to increase the split size by 10% and process all data in one pass. That can improve the read performance up to 50%.

Reply:
>  it would be much better to increase the split size by 10% and process all data in one pass. That can improve the read performance up to 50%.

Maybe the look-ahead can scan the `totalSize` up to 2x of `readToSize`. that might help to make this decision? 

But this batch can always be smaller than the parallelism. we also need change in `planTasksInternal`?

Reply:
I've updated #7714 a bit to cover the two points I mentioned above. I feel using an iterator is a nice solution for regular planning when we don't have a list of tasks. However, it seems redundant for Spark as we keep a list of tasks in memory. I think approached can co-exist.

Reply:
> Suppose we produce 1100 tasks for 1000 cores that are all equally balanced. Spark will run first 1000 tasks in parallel (let's assume each task would take the same amount of time) and only after that schedule the remaining 100 tasks. The remaining 100 tasks would take the same amount of time as the first 1000 as most of the slots in the cluster will be idle. Instead of blindly following the configured split size, it would be much better to increase the split size by 10% and process all data in one pass.

Would it be better? I've typically recommended targeting splits so that you have 2-4x the number of splits as the number of cores. That way if there is a difference in actual task runtime you can still balance the workload. For example, if you're reading 100MB splits where some of them produce 250MB of shuffle data and some produce 2.5MB of shuffle data, you want more splits to balance the workload as executors become idle.

Reply:
Generating 2x by default seems reasonable to me to account for cases when certain tasks finish faster. Whether to use 1x or 2x or 4x would depend on a particular workload. I feel it is important to produce a multiple of the slots in the cluster, however. I am not sure there is a point to aim for 1.5 x splits, for instance. That's what I tried in #7714.

Reply:
I gave a bit of testing to the approach in #7714 and it led to some surprising [results](https://github.com/apache/iceberg/pull/7714#issuecomment-1603503322).